### PR TITLE
Allow models to contain string-based datetime fields that indicate UTC

### LINKED ISSUE #135: Allow models to contain string-based datetime fields that indicate UTC
# Feature request

### Context

I'm attempting to use odmantic with data in an existing Mongo database, but any models of documents containing datetime strings in simplified extended ISO format (i.e., with a trailing "Z") are rejected with a `ValueError`.

Some programming languages (like JavaScript) may express a datetime string [in simplified extended ISO format](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString) ending with a trailing "Z" to indicate UTC.

```javascript
new Date().toISOString() // --> '2021-06-02T02:46:57.366Z'
```

If applications written in those languages write to a Mongo database, datetime strings stored in documents may contain that trailing "Z". Unfortunately, odmantic emits a `ValueError` when it sees the above valid UTC timestamp in any datetime fields of a document.

### Solution

Since [Mongo stores times as UTC](https://docs.mongodb.com/manual/reference/bson-types/#std-label-document-bson-type-date), models should allow string-based datetime fields to specify the time zone or GMT offset as long as the time is expressed as UTC.

For example, the following should be allowed:

* `2018-11-02T23:59:01.824` (assumed to be UTC, currently works)
* `2018-11-02T23:59:01.824Z` (UTC in simplified extended ISO format, currently emits a `ValueError`)
* `2018-11-02T23:59:01.824+00:00` (GMT, equivalent to UTC, currently emits a `ValueError`)

And the following should be disallowed, resulting in a `ValueError`:

* `2018-11-02T23:59:01.824+06:00` (not UTC)

#### Alternative solutions

Ideally, developers would guarantee that any client writing to their Mongo database will store datetime data as a string in naive form or as an actual [timestamp](https://docs.mongodb.com/manual/reference/bson-types/) type, but that's not always possible.

If the data is already in the database, then the other alternative is to modify all existing datetime strings so that they do not end with a "Z" or "+00:00". This shouldn't be necessary and could be cumbersome for large databases.

### PR DESCRIPTION
**Pull request by kfox**
Treats the following as valid datetime string values:

* `2018-11-02T23:59:01.824` (assumed to be UTC)
* `2018-11-02T23:59:01.824Z` (UTC in simplified extended ISO format)
* `2018-11-02T23:59:01.824+00:00` (GMT, equivalent to UTC)

Fixes #135.

### CODE REVIEW

File: tests/unit/test_bson_fields.py [hunk 1]
```diff
@@ -20,12 +20,19 @@ def test_datetime_non_naive():
     class ModelWithDate(Model):
         field: datetime
 
-    with pytest.raises(ValueError):
+    try:
         ModelWithDate(field=datetime.now(tz=pytz.utc))
+        ModelWithDate(field="2018-11-02T23:59:01.824Z")
+        ModelWithDate(field="2018-11-02T23:59:01.824+00:00")
+    except ValueError:
+        pytest.fail("UTC should be acceptable, but is not")
```
Comment (art049) [inline tests/unit/test_bson_fields.py:28]: I think, you can move those to a new test and for example check that the field value is correct for all those

**Review by art049 — CHANGES_REQUESTED** [review]
Hey thanks for the PR ! This didn't come to my attention earlier.
I totally agree. There is just a small thing to change in the test but otherwise everything is good :)

File: tests/unit/test_bson_fields.py (hunk 1 above)
Comment (kfox) [reply tests/unit/test_bson_fields.py:28]: Will do!

**Comment by kfox** [comment]
I refactored the tests. Are any other changes needed?

**Comment by art049** [comment]
Everything looks good! Thank you for the contribution.
Sorry it took that long to merge this PR.

**Pull request merged**